### PR TITLE
Andrews

### PR DESCRIPTION
**Pull request by mcplums**
This adds two things, and their tests:

1) Creates an extra token for 'sponsoring' an event, whereby it is impossible to win. So if there are two outcomes, there will be three tokens. 
2) Modifiers on placeBet function prevent you from betting with an outcome that does not exist (except the new sponsor outcome of course) and if betting with zero dai.

### CODE REVIEW

File: contracts/MBMarket.sol [hunk 1]
```diff
@@ -53,8 +53,8 @@ contract MBMarket is Ownable, Pausable, ReentrancyGuard {
     mapping(uint256 => uint256[]) private betAmountsArray;
     mapping(uint256 => uint256[]) private timestampsArray;
 
-    uint256 public totalBets;
-    uint256 public betsWithdrawn;
+    uint256 public totalBets; // equivilent to 'max bets' goes up with each new bet but does not go down as winnings are withdrawn
```
Comment (gorgos) [inline contracts/MBMarket.sol:56]:
```suggestion
    uint256 public totalBets; // equivalent to 'max bets' goes up with each new bet but does not go down as winnings are withdrawn
```

File: contracts/MBMarket.sol [hunk 2]
```diff
@@ -95,11 +95,14 @@ contract MBMarket is Ownable, Pausable, ReentrancyGuard {
...
         numberOfOutcomes = _outcomeNamesArray.length;
 
-        // create the tokens
+        // create the tokens for each outcome
         for (uint256 i = 0; i < numberOfOutcomes; i++) {
             _createTokenContract(_outcomeNamesArray[i]);
         }
 
+        // create an additional token which can never win to sponsor an event
+        _createTokenContract('Sponsor');
```
Comment (gorgos) [inline contracts/MBMarket.sol:104]: 'Sponsor' would be best as its own constant string variable.

File: contracts/MBMarket.sol [hunk 3]
```diff
@@ -135,7 +138,20 @@ contract MBMarket is Ownable, Pausable, ReentrancyGuard {
...
     ////////////////////////////////////
     modifier checkState(States requiredState) {
-        require(getCurrentState() == requiredState, 'function cannot be called at this time');
+        require(getCurrentState() == requiredState, 'Function cannot be called at this time');
+        _;
+    }
+
+    /// @notice checks the outcome exists
+    modifier outcomeExists(uint256 _outcome) {
+        // + 1 due to sponsor outcome
```
Comment (gorgos) [inline contracts/MBMarket.sol:147]: Cool, so the idea is we can bet on impossible outcomes to increase the winnings? That's how PoolTogether does it as well, they have a lot of money from their own in the contracts but cant win. Good idea. Does it have to be its own token though?

File: contracts/MBMarket.sol [hunk 4]
```diff
@@ -135,7 +138,20 @@ contract MBMarket is Ownable, Pausable, ReentrancyGuard {
...
+        require(_outcome < (numberOfOutcomes + 1), 'Outcome does not exist');
```
Comment (elektronaut0815) [inline contracts/MBMarket.sol:148]: I'd prefer `require(_outcome <= numberOfOutcomes, 'Outcome does not exist');`

File: contracts/MBMarket.sol [hunk 5]
```diff
@@ -247,7 +263,8 @@ contract MBMarket is Ownable, Pausable, ReentrancyGuard {
     function getUserBetsAllOutcomes() public view returns (uint256) {
         // get total bet across all outcomes
         uint256 _userBetsAllOutcomes;
-        for (uint256 i = 0; i < numberOfOutcomes; i++) {
+        for (uint256 i = 0; i < numberOfOutcomes + 1; i++) {
```
Comment (elektronaut0815) [inline contracts/MBMarket.sol:266]: Again prefering `i <= numberOfOutcomes;`

File: contracts/MBMarket.sol [hunk 6]
```diff
@@ -402,7 +426,8 @@ contract MBMarket is Ownable, Pausable, ReentrancyGuard {
     }
 
     function _burnUsersTokens() internal {
-        for (uint256 i = 0; i < numberOfOutcomes; i++) {
+        for (uint256 i = 0; i < numberOfOutcomes + 1; i++) {
```
Comment (elektronaut0815) [inline contracts/MBMarket.sol:429]: Again `i <= numberOfOutcomes;`

Comment (mcplums) [reply contracts/MBMarket.sol:429]: I've made this change

File: contracts/MBMarket.sol (hunk 4 above)
Comment (mcplums) [reply contracts/MBMarket.sol:148]: I've made this change

File: contracts/MBMarket.sol (hunk 5 above)
Comment (mcplums) [reply contracts/MBMarket.sol:266]: I've made this change

**Comment by mcplums** [comment]
I've fixed the coding style thing. But I think we should leave the token because it means sponsors can use the existing placeBet() and withdraw() functions. Without a token we would need to create new functions which would just make things needlessly complicated. What do you think?

**Comment by mcplums** [comment]
Ok, so basically, removing the creation of the sponsor token from the constructor to this new function?

**Comment by gorgos** [comment]
> Ok, so basically, moving the creation of the sponsor token from the constructor to this new function?

No token at all, just something like

```
function sponsorEvent(uint256 sponsorAmount) external onlyOwner {
    dai.transferFrom(msg.sender, address(this), sponsorAmount);
    sponsoredMoney = sponsoredMoney.add(sponsorAmount);
}
```

**Comment by mcplums** [comment]
Ok- I was about to say this wouldn't be cleaner because we'd also need a new withdraw function (i.e. two new functions when zero needed if we just add a function) but we would probably need a new withdraw function anyway. The current withdraw function can only be used in 'withdraw' state, but the sponsor should be able to withdraw whenever they want. 

Only thing I don't understand- what's the reason to make it onlyOwner and have a corresponding function in the factory?

**Comment by gorgos** [comment]
> Only thing I don't understand- what's the reason to make it onlyOwner and have a corresponding function in the factory?

Optional, just to prevent dumb users from calling this, but I guess we dont have to make it onlyOwner.

**Comment by mcplums** [comment]
But if its onlyOwner only we can call it- the function is there not so much for us, but for places like Coinbase to add funds :)

I don't think it matters if anyone calls it accidentally, they can always get it back again immediately.

**Comment by mcplums** [comment]
I don't agree (or understand) your point about onlyOwner. It would not work with onlyOwner. So I have not added this. I also no longer think it makes sense to add a separate withdraw function for the sponsor, it is not a problem for them to have to wait until the market resolves. 

Given this, I think it is ready to merge now? There are no changes at all.

**Comment by mcplums** [comment]
I think a sponsor token is the best solution for two reasons:
1) Can use existing code instead of having to create new functions
2) Something tangible for sponsors to see in return for their sponsorship. 

But I don't have a strong opinion, if you feel strongly that it should not be a token, please code that up.

**Comment by mcplums** [comment]
I notice the tests fail, I will fix this tomorrow :)

**Comment by gorgos** [comment]
I think we can close this now due to #44

**Pull request closed without merging**